File: GeoServices/SDE/EditingWorkspace.vb
Imports ESRI.ArcGIS.Geodatabase

Namespace SDE
    Public Class EditingWorkspace
        Implements IEditingObject

        Protected Property Workspace As IWorkspace
        Protected ReadOnly Property EditWorkspace As IWorkspaceEdit2
            Get
                Return CType(Me.Workspace, IWorkspaceEdit2)
            End Get
        End Property

        Public Sub New(ByVal workspace As IWorkspace)
            If Not TypeOf workspace Is IWorkspaceEdit2 Then Throw New DataException("El workspace no es válido para edición")
            Me.Workspace = workspace
        End Sub

        Public Function isVersioned() As Boolean Implements SDE.IEditingObject.isVersioned
            Dim version As String = Me.Workspace.ConnectionProperties.GetProperty("VERSION")
            Return Not version Is Nothing AndAlso Not version = ""
        End Function

        Public Sub StartEditing(ByVal WithUndoRedo As Boolean) Implements SDE.IEditingObject.StartEditing
            If Me.isVersioned() Then Me.EditWorkspace.StartEditing(WithUndoRedo)
        End Sub

        Public Sub StartEditOperation() Implements SDE.IEditingObject.StartEditOperation
            If Me.isVersioned() Then Me.EditWorkspace.StartEditOperation()
        End Sub

        Public Sub StopEditOperation() Implements SDE.IEditingObject.StopEditOperation
            If Me.isVersioned() Then Me.EditWorkspace.StopEditOperation()
        End Sub

        Public Sub StopEditing(ByVal SaveEdits As Boolean) Implements SDE.IEditingObject.StopEditing
            If Me.isVersioned() Then Me.EditWorkspace.StopEditing(SaveEdits)
        End Sub
    End Class
End Namespace

File: GeoServicesSharp/SDE/EditingWorkspace.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using ESRI.ArcGIS.Geodatabase;
using GeoServicesSharp.Exceptions;

namespace GeoServicesSharp.SDE
{
    class EditingWorkspace : IEditingObject
    {
        protected IWorkspace Workspace {get; set;}
        protected IWorkspaceEdit2 EditWorkspace
        {
            get
            {
                return (IWorkspaceEdit2)this.Workspace;
            }
        }

        public EditingWorkspace(IWorkspace workspace)
        {
            if (!(workspace is IWorkspaceEdit2)) throw new GeoServicesException("El workspace no es válido para edición");
            this.Workspace = workspace;
        }

        public bool isVersioned()
        {
            string version = (string)this.Workspace.ConnectionProperties.GetProperty("VERSION");
            return !(version == null) &&  !(version == "");
        }

        public void StartEditing(bool WithUndoRedo)
        {
            if (this.isVersioned()) this.EditWorkspace.StartEditing(WithUndoRedo);
        }

        public void StartEditOperation()
        {
            if (this.isVersioned()) this.EditWorkspace.StartEditOperation();
        }

        public void StopEditOperation()
        {
            if (this.isVersioned()) this.EditWorkspace.StopEditOperation();
        }

        public void StopEditing(bool SaveEdits)
        {
            if (this.isVersioned()) this.EditWorkspace.StopEditing(SaveEdits);
        }
    }
}
